src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -1197,4 +1346,101 @@ ObjectDataHandler::clear_ret ObjectDataHandler::clear(
     });
 }
 
+ObjectDataHandler::clone_ret ObjectDataHandler::clone(
+  context_t ctx)
+{
+  return with_objects_data(
+    ctx,
+    [ctx, this](auto &object_data, auto &d_object_data) {
+    LOG_PREFIX(ObjectDataHandler::clone);
+    ceph_assert(d_object_data.is_null());
```

Comment:
The issue was as follows:
If we let `object_data` hold the original data mappings and let `d_object_data` hold the indirect mappings, we'll have modify all the indirect mappings held by all clones. This is not just a performance issue, as we don't know which clones are holding those to-be-modified indirect mappings in seastore when writing data to the head object. So I reversed the clone direction.

The problem with the reversed clone direction is that very read issued to an not-modified extent has to go through two lba searches, one to find the indirect mapping, the other to find the original data mapping. But this can be solved in this way: with the previously discussed [`lba + 1` entries](https://docs.google.com/document/d/1TOZ1jnlxIKwmp3CE7Vr7vE1vujergXa9/edit#) in place, we can always let head extents has two lba mappings in the lba tree, one points to the actual data, the other holds the indirect mappings; when doing overwrite, we drop the indirect mappings, while when doing clone, we drop the actual data mapping. One thing to note is that this may mean that the backref tree needs to be touched eventually.

Reply:
> If we let object_data hold the original data mappings and let d_object_data hold the indirect mappings, we'll have modify all the indirect mappings held by all clones when overwriting an extent.

This looks to have the assumption that `d(est)_object_data` will have more chances to be modified than the cloned `object_data`. So the decision here is to make `d(est)_object_data` to take the direct mappings. I'm not sure about the specific scenario from OSD perspective, @athanatos do you have more insights?

> This is not just a performance issue, as we don't know which clones are holding those to-be-modified indirect mappings in seastore when writing data to the head object.

This looks to be a correctness problem, can you elaborate more about why `we don't know which clones are holding those to-be-modified indirect mappings` is a problem, probably with an example?

> The problem with the reversed clone direction is that every read issued to an non-modified extent has to go through two lba searches, one to find the indirect mapping, the other to find the original data mapping.

This is a performance consideration that can be postponed. However, to extend a little bit, compared with `lba + 1` solution, seems building up a laddr -> CachedExtent LRU cache layer would be more efficient because there this is a shorter/denser lookup path, and there is no requirement to make this persistent. We can discuss more later on.

Reply:
> This looks to have the assumption that `d(est)_object_data` will have more chances to be modified than the cloned `object_data`. So the decision here is to make `d(est)_object_data` to take the direct mappings. I'm not sure about the specific scenario from OSD perspective, @athanatos do you have more insights?
> 
> >

>This looks to be a correctness problem, can you elaborate more about why we don't know which clones are holding those to-be-modified indirect mappings is a problem, probably with an example?

Clones doesn't get modified. The problem is as follows:

1. Head objects has an extent `E`;
2. Clone1..N objects have indirect mappings pointing to `E`;
3. An overwrite is issued to `E`, we'll have to modify the laddr of `E` because the original laddr need to point to the new version of `E`, this requires that all indirect mappings pointing to `E` to be modified;

This is problematic because we don't know anything about clone 1...N when overwriting head objects

Reply:
> This is a performance consideration that can be postponed. However, to extend a little bit, compared with `lba + 1` solution, seems building up a laddr -> CachedExtent LRU cache layer would be more efficient because there this is a shorter/denser lookup path, and there is no requirement to make this persistent. We can discuss more later on.


The lba tree with pointers in #47749 seems to the `laddr -> CachedExtent LRU cache layer`, am I right?

Reply:
> 3. An overwrite is issued to E, we'll have to modify the laddr of E because the original laddr need to point to the new version of E, this requires that all indirect mappings pointing to E to be modified;

Ah, I see this is because we statically map object-offset to lba, so if that static lba has to be mapped to the new verson of the data extent upon modification, it can break other clones if it is indirectly pointed (ref-count > 1).

We really should assert in this case that any lba that has ref-counter larger than 1 should not be remapped to a different extent.

Reply:
> We really should generally assert in this case that any lba that has ref-counter larger than 1 should not be remapped to a different extent.

Yeah, will add it.

Reply:
> The lba tree with pointers in https://github.com/ceph/ceph/pull/47749 seems to the laddr -> CachedExtent LRU cache layer, am I right?

Yeah, but the lba tree implies the mappings needs to be persistent (with overhead), and the B+ tree is optimized for both block and memory which has to operate at the block granularity.

I think if laddr translation is really performance critical, probably it finally worths to experiment and introduce a pure volatile LRU cache layer (laddr -> CachedExtents) which works above the lba tree at runtime.

Reply:
> (off-topic)

Yes, this really is off-topic, but please allow me to discuss this a little bit more since I can't find an appropriate place to talk about this.

> 
> Yeah, but the lba tree implies the mappings needs to be persistent (with overhead),

I understand your point. But I think, whether we use the `laddr -> CachedExtents LRU`  or not, we always have to persist the information about those `extra laddr -> CachedExtents' paddr` to disk. It's just that it's not persisted by the `laddr -> CachedExtents LRU` but by the lba tree.

Indeed, the `lba + 1` approach has to persist an extra lba mapping for each extent that's shared by the head object and the clone object; but, on the other hand, the `laddr -> CachedExtents LRU` probably will end up duplicating the whole of the in-memory part of the lba tree, which seems to me to be a bigger overhead.

> and the B+ tree is optimized for both block and memory which has to operate at the block granularity.
> 
> I think if laddr translation is really performance critical, probably it finally worths to experiment and introduce a pure volatile LRU cache layer (laddr -> CachedExtents, no intermediate laddr -> laddr) which works above the lba tree at runtime.

With the `lba + 1` approach, only clone reads may need to follow the intermediate laddr->laddr, if the extents accessed is an indirect one; access to head objects won't need to do that.

Reply:
> Yes, this really is off-topic, but please allow me to discuss this a little bit more since I can't find an appropriate place to talk about this.

Ah no warries, I just want to label out the independent discussion thread in case of interleaving.

> But I think, whether we use the laddr -> CachedExtents LRU or not, we always have to persist the information about those extra laddr -> CachedExtents' paddr mappings to disk.

Hmm, this looks to be a reasonable point to make it persistent.

Reply:
@xxhdx1985126 @athanatos The agreement is to implement symmetric clones, so that both oid and noid will use their own "actual" laddrs pointing to a shared "indirect" laddr. Also, try to make the initial version simple without subtle optimizations.

Let's formalize the terms in this PR about the actual or indirect laddr (or pick different names as long as they can be consistently used in code), and hide the special indirect laddr in the lba tree wherever possible.

Another related comment here is: Let's use the original `prepare_data_reservation()` like in other places, which is only to make sure the nullable `object_data` (not d_object_data) is properly initialized before any followup clone operations. I think this might be more understandable.

Does it make sense?

Reply:
Ok